File: extension/background/intentParser.js
/* globals log */

import { PhraseSet } from "./language/matching.js";
import { compile, splitPhraseLines } from "./language/compiler.js";
import { entityTypes } from "./entityTypes.js";

const DEFAULT_INTENT = "search.search";
const DEFAULT_SLOT = "query";

// Populated by registerMatcher:
const INTENTS = {};
const INTENT_NAMES = [];

export function registerMatcher(intentName, matcher) {
  if (initialized) {
    throw new Error(`Late attempt to register intent: ${intentName}`);
  }
  if (INTENTS[intentName]) {
    throw new Error(`Intent ${intentName} has already been registered`);
  }
  INTENT_NAMES.push(intentName);
  INTENT_NAMES.sort();
  INTENTS[intentName] = { matcher };
}

let initialized = false;
let phraseSet;

function initialize() {
  const phrases = [];
  for (const name in INTENTS) {
    const { matcher } = INTENTS[name];
    for (const line of splitPhraseLines(matcher)) {
      const compiled = compile(line, {
        entities: entityTypes,
        intentName: name,
      });
      INTENTS[name].compiledMatcher = compiled;
      phrases.push(compiled);
    }
  }
  phraseSet = new PhraseSet(phrases);
  initialized = true;
}

export function parse(text, disableFallback = false) {
  if (!initialized) {
    initialize();
  }
  const result = phraseSet.match(text);
  if (!result) {
    if (disableFallback) {
      return null;
    }
    log.info(`Parsed as fallback intent: ${JSON.stringify(text)}`);
    return {
      name: DEFAULT_INTENT,
      slots: { [DEFAULT_SLOT]: text },
      parameters: {},
      utterance: text,
      fallback: true,
    };
  }
  return {
    name: result.intentName,
    slots: result.stringSlots(),
    parameters: result.parameters,
    utterance: text,
    fallback: false,
  };
}

export function getIntentNames() {
  return INTENT_NAMES;
}

File: app/src/main/java/mozilla/voice/assistant/IntentParser.kt
package mozilla.voice.assistant

import androidx.annotation.VisibleForTesting
import mozilla.voice.assistant.language.Compiler
import mozilla.voice.assistant.language.Pattern
import mozilla.voice.assistant.language.PhraseSet

fun <T> Map<T, T>.getOrKey(key: T): T = this[key] ?: key

class IntentParser {
    companion object {
        private var initialized = false
        private val compiledPhrases = mutableListOf<Pattern>()
        private lateinit var phraseSet: PhraseSet

        // Populated by registerMatcher
        @VisibleForTesting
        private val INTENT_NAMES = mutableListOf<String>() // all caps to match JS code

        private const val DEFAULT_INTENT = "search.search"
        private const val DEFAULT_SLOT = "query"

        internal fun registerMatcher(intentName: String, match: List<String>) {
            if (initialized) {
                throw Error("Late attempt to register intent $intentName")
            }
            if (INTENT_NAMES.contains(intentName)) {
                throw Error("Intent $intentName has already been registered")
            }
            INTENT_NAMES.add(intentName)
            match.forEach {
                // TODO: Add entity names
                compiledPhrases.add(Compiler.compile(it, emptyMap(), intentName))
            }
        }

        internal fun initialize() {
            require(!initialized)
            phraseSet = PhraseSet(compiledPhrases)
            initialized = true
        }

        internal fun getIntentNames() = INTENT_NAMES

        internal fun parse(text: String, disableFallback: Boolean = false): ParseResult? {
            require(initialized)
            val matchResult = phraseSet.match(text)
            if (matchResult == null) {
                return if (disableFallback) {
                    null
                } else {
                    ParseResult(
                        name = DEFAULT_INTENT,
                        slots = mapOf(DEFAULT_SLOT to text),
                        utterance = text,
                        fallback = true
                    )
                }
            }
            return ParseResult(
                name = matchResult.intentName ?: throw error("MatchResult in parse() lacks intentName"),
                slots = matchResult.stringSlots(),
                parameters = matchResult.parameters,
                utterance = text,
                fallback = false
            )
        }
    }
}

class ParseResult(
    val utterance: String,
    val name: String,
    val slots: Map<String, String>,
    val parameters: Map<String, String> = emptyMap(),
    val fallback: Boolean
)
